CAIPs/caip-x.md

```diff
@@ -0,0 +1,326 @@
+
+## Specification
+
+### New Methods
+
+#### `provider_augment`
+
+Adds new permissions to an existing session.
+
+**Parameters:**
+
+- `sessionId` (string, optional): The session identifier.
+- `scopes` (object, required): An object containing the new scopes to be added, formatted according to CAIP-217.
+
+**Initial Session Scopes:**
+
+```json
+{
+  "eip155:1": {
+    "methods": ["eth_signTransaction"],
+    "notifications": ["accountsChanged"]
+  },
+  "eip155:137": {
+    "methods": ["eth_sendTransaction"],
+    "notifications": ["chainChanged"]
+  }
+}
+```
+
+**Example Request:**
+
+```json
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "method": "provider_augment",
+  "params": {
+    "scopes": {
+      "eip155:1": {
+        "methods": ["eth_sendTransaction"],
+        "notifications": ["chainChanged"]
+      }
+    }
+  }
+}
+```
+
+**Example Response:**
+
+```json
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "result": {
+    "success": true,
+    "updatedScopes": {
+      "eip155:1": {
+        "methods": ["eth_signTransaction", "eth_sendTransaction"],
+        "notifications": ["accountsChanged", "chainChanged"]
+      },
+      "eip155:137": {
+        "methods": ["eth_sendTransaction"],
+        "notifications": ["chainChanged"]
+      }
+    }
+  }
+}
+```
+
+**Explanation:**
+
+- The `provider_augment` method adds the specified scopes to the current session's permissions. If the scope already exists, the new methods and notifications are merged with the existing ones.
+
+#### `provider_revoke`
+
+Revokes permissions from an existing session.
+
+**Parameters:**
+
+- `sessionId` (string, optional): The session identifier.
+- `scopes` (object, required): An object containing the scopes to be revoked, formatted according to CAIP-217.
+
+**Initial Session Scopes:**
+
+```json
+{
+  "eip155:1": {
+    "methods": ["eth_signTransaction", "eth_sendTransaction"],
+    "notifications": ["accountsChanged", "chainChanged"]
+  },
+  "eip155:137": {
+    "methods": ["eth_sendTransaction"],
+    "notifications": ["chainChanged"]
+  }
+}
+```
+
+**Example Request:**
+
+```json
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "method": "provider_revoke",
+  "params": {
+    "scopes": {
+      "eip155:1": {
+        "methods": ["eth_sendTransaction"],
+        "notifications": ["chainChanged"]
+      }
+    }
+  }
+}
+```
+
+**Example Response:**
+
+```json
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "result": {
+    "success": true,
+    "updatedScopes": {
+      "eip155:1": {
+        "methods": ["eth_signTransaction"],
+        "notifications": ["accountsChanged"]
+      },
+      "eip155:137": {
+        "methods": ["eth_sendTransaction"],
+        "notifications": ["chainChanged"]
+      }
+    }
+  }
+}
+```
+
+**Explanation:**
+
+- The `provider_revoke` method removes the specified methods and notifications from the current session's permissions. If the scope no longer has any methods or notifications after revocation, it is removed from the session.
```

Comment:
what if a session has no scopes left?

Reply:
Yeah we should call that out. I suppose the "session" is closed.

Reply:
they could revoke `account` permissions in the session as well, right?

Reply:
yes for sure

Reply:
added some language to account for this and added `accounts` to the examples: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/51721fd888285fe5bba9bd3ca6d98929c73c5772

Reply:
I think if no scopes are provided the method should treat it as a revocation of the whole session. No need to send the whole permission object over the wire if you can express intent in a more efficient manner. We can also of course continue to allow a caller to list the whole session if they so wish...

Reply:
Yeah that makes sense to me. Perhaps in the same vein sending an empty scopeObject revokes everything for that scope?